### ISSUE TITLE
RFC: Eliminate per-context custom allocators

### ISSUE THREAD

**Issue by nathanaeljones** [imazen/imageflow#82]
Rust does not currently support custom allocators except in its nightly build - and for the entire executable at that. 

Our C code currently checks all allocations for failure, and gracefully returns; every method has control flow to deal with a malloc failure at any point.

In the past, we used imageflow's  custom allocators to fully exercise those code paths. In the future, we may have to swap out the entire allocator for the test process, and use global state to simulate failures. This is less clean, but we benefit from Rust's safety guarantees. 
## How do you handle malloc failure in rust?

Well, [Rust can now catch panics and handle them gracefully](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html), so we don't have to compromise on our C ABI's behavior.

That said, overcommit is a thing (if [not to the degree people assume](https://www.etalabs.net/overcommit.html)), so most software just crashes [when malloc fails](http://eli.thegreenplace.net/2009/10/30/handling-out-of-memory-conditions-in-c).

We probably can't do anything about process failure when we try to access memory that malloc provided, yet can't actually be accessed. We _can_ handle exhaustion of the virtual memory space. 

We also know which allocations will be huge, and could directly handle those allocations (we don't even need panic/catch for those). 

**To sum up, the only sacrifice we would need to make is dropping support for custom allocators per context. As some codecs didn't permit injecting alternate allocators, this was not a promise we could ever completely fulfill).** 

And that doesn't mean we can't use custom allocators for the largest allocations; just not for tiny allocations that support non-pixel data.

**Comment by nathanaeljones** [comment]
Development will proceed as if this change is acceptable. If there are hard requirements around this, please bring them up ASAP.

**Comment by nathanaeljones** [comment]
No more custom allocators. You can use whichever allocators you want via Rust's API, but it's process-wide.
